rocAL/source/loaders/audio/audio_loader.cpp

```diff
@@ -0,0 +1,288 @@
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+*/
+
+#include <thread>
+#include <chrono>
+#include "audio_loader.h"
+#include "audio_read_and_decode.h"
+#include "vx_ext_amd.h"
+
+AudioLoader::AudioLoader(void* dev_resources):
+_circ_buff(dev_resources),
+_swap_handle_time("Swap_handle_time", DBG_TIMING) {
+    _output_tensor = nullptr;
+    _mem_type = RocalMemType::HOST;
+    _internal_thread_running = false;
+    _output_mem_size = 0;
+    _batch_size = 1;
+    _is_initialized = false;
+    _remaining_audio_count = 0;
+    _device_id = 0;
+}
+
+AudioLoader::~AudioLoader() {
+    de_init();
+}
+
+void AudioLoader::shut_down() {
+    if(_internal_thread_running)
+        stop_internal_thread();
+    _circ_buff.release();
+}
+
+void AudioLoader::set_prefetch_queue_depth(size_t prefetch_queue_depth) {
+    if(prefetch_queue_depth <= 0)
+        THROW("Prefetch queue depth value cannot be zero or negative");
+    _prefetch_queue_depth = prefetch_queue_depth;
+}
+
+void AudioLoader::set_gpu_device_id(int device_id) {
+    if(device_id < 0)
+        THROW("invalid device_id passed to loader");
+    _device_id = device_id;
+}
+
+size_t
+AudioLoader::remaining_count() {
+    return _remaining_audio_count;
+}
+
+void AudioLoader::reset() {
+    // stop the writer thread and empty the internal circular buffer
+    _internal_thread_running = false;
+    _circ_buff.unblock_writer();
+    if (_load_thread.joinable())
+        _load_thread.join();
+    // Emptying the internal circular buffer
+    _circ_buff.reset();
+    // resetting the reader thread to the start of the media
+    _audio_counter = 0;
+    _audio_loader->reset();
+    // Start loading (writer thread) again
+    start_loading();
+}
+
+void AudioLoader::de_init() {
+    // Set running to 0 and wait for the internal thread to join
+    stop_internal_thread();
+    _output_mem_size = 0;
+    _batch_size = 1;
+    _is_initialized = false;
+}
+
+LoaderModuleStatus
+AudioLoader::load_next() {
+    return update_output_audio();
+}
+
+void AudioLoader::set_output(Tensor* output_tensor) {
+    _output_tensor = output_tensor;
+    _output_mem_size = _output_tensor->info().data_size();
+}
+
+void AudioLoader::stop_internal_thread() {
+    _internal_thread_running = false;
+    _stopped = true;
+    _circ_buff.unblock_reader();
+    _circ_buff.unblock_writer();
+    _circ_buff.reset();
+    if (_load_thread.joinable())
+        _load_thread.join();
+}
+
+void AudioLoader::initialize(ReaderConfig reader_cfg, DecoderConfig decoder_cfg, RocalMemType mem_type, unsigned batch_size, bool decoder_keep_original) {
```

Comment:
decode_keep_original is not used, why is it introduced as arg? Please check

Reply:
Its a pure virtual function in the base class